src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -25,16 +35,121 @@
 
   /*
    * Called when a client submits a GET request to the /data URL
-   * Displays all recorded user comments on page
    */
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {}
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // PROTO Data structure:
+    // Parse the contents  of graph.txt into a proto Graph object, and extract information
+    // from the proto object into a map. This is used to store the proto input and isn't updated
+    // with mutations.
+    Graph protoGraph =
+        Graph.parseFrom(getServletContext().getResourceAsStream("/WEB-INF/graph.txt"));
+    Map<String, Node> protoNodesMap = protoGraph.getNodesMapMap();
+
+    // GRAPH Data structures:
+    // Create an undirected graph data structure to store the information, and
+    // map each node name in the graph to the GraphNode objects. This is the graph & map
+    // we update with mutations
+    MutableGraph<GraphNode> graph = GraphBuilder.directed().build();
+    HashMap<String, GraphNode> graphNodesMap = new HashMap<>();
+
+    // Create a graph (useable form) from proto data structures
+    for (String nodeName : protoNodesMap.keySet()) {
+
+      Node thisNode = protoNodesMap.get(nodeName);
+
+      // Convert thisNode into a graph node that may store additional information
+      GraphNode graphNode = protoNodeToGraphNode(thisNode);
+
+      // Update graph data structures to include the node
+      graph.addNode(graphNode);
+      graphNodesMap.put(nodeName, graphNode);
+
+      // Add dependency edges to the graph
+      for (String child : thisNode.getChildrenList()) {
+        graph.putEdge(graphNode, protoNodeToGraphNode(protoNodesMap.get(child)));
+      }
+    }
+
+    // Parse the contents of mutation.txt into a list of mutations
+    List<Mutation> mutList =
+        MutationList.parseFrom(getServletContext().getResourceAsStream("/WEB-INF/mutations.txt"))
+            .getMutationList();
+
+    for (Mutation mut : mutList) {
+      // Nodes affected by the mutation
+      // second node only applicable for adding an edge and removing an edge
+      String startName = mut.getStartNode();
+      String endName = mut.getEndNode();
+
+      // Getting the corresponding graph nodes from the graph map
+      GraphNode startNode = graphNodesMap.get(startName);
+      GraphNode endNode = graphNodesMap.get(endName);
+
+      switch (mut.getType()) {
+        case ADD_NODE:
+          // Create a new node with the given name and add it to the graph and the map
+          Node newNode = Node.newBuilder().setName(startName).build();
+          GraphNode newGraphNode = protoNodeToGraphNode(newNode);
```

Comment:
Parsing the .proto file gives us a Graph object which contains a Map from string name to Node. We created a separate graph node because we thought that in the future we might add attributes like changedFromPrev graph (for color-coding the diffs) to the node but we couldn't modify the proto to get that in the Node object.

Reply:
What I meant was why not just do something like `newGraphNode = GraphNode.create(...)` here, but I guess converting it to a proto Node first handles the default values for token_list and metadata :)

Reply:
Yup exactly :)